Firebase/Storage/FIRStorageListTask.m

```diff
@@ -0,0 +1,127 @@
+  NSString *_previousPageToken;
+}
+
+@synthesize fetcher = _fetcher;
+@synthesize fetcherCompletion = _fetcherCompletion;
+
+- (instancetype)initWithReference:(FIRStorageReference *)reference
+                   fetcherService:(GTMSessionFetcherService *)service
+                    dispatchQueue:(dispatch_queue_t)queue
+                         pageSize:(nullable NSNumber *)pageSize
+                previousPageToken:(nullable NSString *)previousPageToken
+                       completion:(FIRStorageVoidListError)completion {
+  self = [super initWithReference:reference fetcherService:service dispatchQueue:queue];
+  if (self) {
+    _completion = [completion copy];
+    _pageSize = pageSize;
+    _previousPageToken = previousPageToken;
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [_fetcher stopFetching];
+}
+
+- (void)enqueue {
+  __weak FIRStorageListTask *weakSelf = self;
+
+  [self dispatchAsync:^() {
+    FIRStorageListTask *strongSelf = weakSelf;
+
+    if (!strongSelf) {
+      return;
+    }
+
+    NSMutableDictionary<NSString *, NSString *> *queryParams = [NSMutableDictionary new];
+
+    NSString *prefix = [self reference].fullPath;
+    if (prefix.length != 0) {
+      queryParams[@"prefix"] = [prefix stringByAppendingString:@"/"];
+    }
+
+    // Firebase Storage uses file system semantics and treats slashes as separators. GCS's List API
+    // does not prescribe a separator, and hence we need to provide a slash as the delimiter.
+    queryParams[@"delimiter"] = @"/";
+
+    // listAll() doesn't set a pageSize as this allows Firebase Storage to determine how many items
+    // to return per page. This removes the need to backfill results if Firebase Storage filters
+    // objects that are considered invalid (such as items with two consecutive slashes).
+    if (strongSelf->_pageSize) {
+      queryParams[@"maxResults"] = [strongSelf->_pageSize stringValue];
+    }
+
+    if (strongSelf->_previousPageToken) {
+      queryParams[@"pageToken"] = strongSelf->_previousPageToken;
+    }
+
+    FIRStoragePath *basePath = [[FIRStoragePath alloc] initWithBucket:self.reference.bucket
+                                                               object:nil];
+    NSMutableURLRequest *request =
+        [[FIRStorageUtils defaultRequestForPath:basePath queryParams:queryParams] mutableCopy];
+
+    request.HTTPMethod = @"GET";
+    request.timeoutInterval = strongSelf.reference.storage.maxOperationRetryTime;
+
+    FIRStorageVoidListError callback = strongSelf->_completion;
+    strongSelf->_completion = nil;
+
+    GTMSessionFetcher *fetcher = [strongSelf.fetcherService fetcherWithRequest:request];
+    strongSelf->_fetcher = fetcher;
+    fetcher.comment = @"ListTask";
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-retain-cycles"
+    strongSelf->_fetcherCompletion = ^(NSData *data, NSError *error) {
+      FIRStorageListResult *listResult;
+      if (error) {
+        if (!self.error) {
```

Comment:
Is `self` used here intentionally? If yes, a comment explaining why would be helpful.

Reply:
I added the following comment:

"// In case _fetcherCompletion gets called multiple times, don't override a previous error."

The logic is copied from the existing task implementations... Is the comment accurate, Mr. @mcdonamp?

Reply:
> The logic is copied from the existing task implementations...

I see... I suspect, `self` is captured by the block here to keep the task "alive" until completion, then the retain cycle is broken by `self->_fetcherCompletion = nil;`. I hoped to get confirmation on it :) 
Nit: It would be nice to add a short explanation eventually.

Reply:
For `self->_fetcherCompletion = nil` that is true. I added a comment in that line.

I am not 100% certain on why we guard self.error though.

Reply:
No on remembers why we have the `if(self.error)` check so I removed it, rather than continue with the pattern.